apps/chef_objects/src/chef_object.erl

```diff
@@ -207,71 +208,42 @@ fetch(Rec, CallbackFun) ->
     Mod:fetch(Rec, CallbackFun).
 
-fetch_multi(RecModule, QueryName, QueryParams, CallbackFun) ->
+fetch_multi(ApiVersion, RecModule, QueryName, QueryParams, CallbackFun) ->
     CallbackFun({QueryName,
                  QueryParams,
-                 {rows_as_records, [RecModule, RecModule:record_fields()]}}).
+                 {rows_as_records, [RecModule, RecModule:record_fields(ApiVersion)]}}).
 
 update(Rec, ActorId, CallbackFun) ->
     Mod = element(1, Rec),
     Mod:update(chef_object:set_updated(Rec, ActorId), CallbackFun).
 
 delete(Rec, CallbackFun) ->
-    call_if_exported(Rec, delete, [Rec, CallbackFun], fun do_delete/2).
-
-%% Return the callback module for a given object record type. We're putting the abstraction
-%% in place in case we need to do something other than the identity mapping of record name
-%% to callback module name that we're doing there. If we needed to swap in something else,
-%% we could do that mapping here.
-callback_mod(Rec) ->
-    element(1, Rec).
+    call_if_exported(Rec, delete, [Rec, CallbackFun], fun chef_object_default_callbacks:delete/2).
 
 call(Rec, Fun) ->
-    Mod = callback_mod(Rec),
+    Mod = element(1, Rec),
     Mod:Fun(Rec).
 
-call0(Rec, Fun) ->
-    Mod = callback_mod(Rec),
-    Mod:Fun().
 
-
-default_fetch(Rec, CallbackFun) ->
-    CallbackFun({call0(Rec, find_query),
-            call(Rec, fields_for_fetch),
-            {first_as_record, [element(1, Rec), call0(Rec, record_fields)]}}).
-
-default_update(ObjectRec, CallbackFun) ->
-    QueryName = chef_object:update_query(ObjectRec),
-    UpdatedFields = chef_object:fields_for_update(ObjectRec),
-    CallbackFun({QueryName, UpdatedFields}).
-
-flatten(ObjectRec) ->
-    call_if_exported(ObjectRec, flatten, [ObjectRec], fun default_flatten/1).
+fields_for_insert(ObjectRec) ->
+    call_if_exported(ObjectRec, fields_for_insert, [ObjectRec],
+                     fun chef_object_default_callbacks:fields_for_insert/1).
 
 call_if_exported(ObjectRec, FunName, Args, DefaultFun) ->
-    Mod = callback_mod(ObjectRec),
+    Mod = element(1, ObjectRec),
     case erlang:function_exported(Mod, FunName, length(Args)) of
         true ->
             erlang:apply(Mod, FunName, Args);
         false  ->
             erlang:apply(DefaultFun, Args)
     end.
-default_flatten(ObjectRec) ->
-    [_RecName|Tail] = tuple_to_list(ObjectRec),
-    %% We detect if any of the fields in the record have not been set
-    %% and throw an error
-    case lists:any(fun is_undefined/1, Tail) of
-        true -> error({undefined_in_record, ObjectRec});
-        false -> ok
-    end,
-    Tail.
-
-is_undefined(undefined) ->
-    true;
-is_undefined(_) ->
-    false.
-
-do_delete(ObjectRec, CallbackFun) ->
-    QueryName = delete_query(ObjectRec),
-    Id = id(ObjectRec),
-    CallbackFun({QueryName, [Id]}).
+
+set_api_version(Record, ApiVersion) when is_tuple(Record) ->
+    case erlang:tuple_to_list(Record) of
+        [Type, undefined | Rest] ->
```

Comment:
agreed, I know this one had a bit of smell to it even as I did it. Changed it to require a callback behavior from chef_object modules, and use that instead.